src/internet_identity/src/account_management.rs

```diff
@@ -39,6 +44,108 @@ pub fn get_accounts_for_origin(
     storage_borrow(|storage| storage.list_accounts(anchor_number, origin))
 }
 
+/// Helper function to read an account by application number, unlike `storage.read_account` this
+/// returns `Result` instead of `Option`.
+///
+/// This function is applicable only to numbered accounts; synthetic accounts are not currently
+/// stored in the memory and thus cannot be fetched from the storage.
+fn try_read_account_info(
+    anchor_number: AnchorNumber,
+    origin: FrontendHostname,
+    application_number: ApplicationNumber,
+    account_number: AccountNumber,
+) -> Result<AccountInfo, String> {
+    let Some(account) = storage_borrow(|storage| {
+        storage.read_account(ReadAccountParams {
+            account_number: Some(account_number),
+            anchor_number,
+            origin: &origin,
+            known_app_num: Some(application_number),
+        })
+    }) else {
+        let message = format!(
+            "Account #{} does not exist for anchor {} and origin {}.",
+            account_number, anchor_number, origin
+        );
+
+        return Err(message);
+    };
+
+    Ok(account.to_info())
+}
+
+/// Best effort to determin the default account for the given (anchor, origin).
+///
+/// An Err case would indicate internal inconsistency in the canister state.
+pub fn get_default_account_for_origin(
+    anchor_number: AnchorNumber,
+    origin: FrontendHostname,
+) -> Result<AccountInfo, GetDefaultAccountError> {
+    let Some(application_number) =
+        storage_borrow(|storage| storage.lookup_application_number_with_origin(&origin))
+    else {
+        return Ok(Account::synthetic(anchor_number, origin).to_info());
+    };
+
+    let AnchorApplicationConfig {
+        default_account_number,
+    } = storage_borrow(|storage| {
+        storage.lookup_anchor_application_config(anchor_number, application_number)
+    });
+
+    let Some(default_account_number) = default_account_number else {
+        return Ok(Account::synthetic(anchor_number, origin).to_info());
+    };
+
+    let account = try_read_account_info(
+        anchor_number,
+        origin,
+        application_number,
+        default_account_number,
+    )
+    .map_err(GetDefaultAccountError::InternalCanisterError)?;
+
+    Ok(account)
+}
+
+/// Sets the default account for the given (anchor, origin) to the specified `account_number`.
+///
```

Comment:
I would add a comment that if the account_number is None, then the synthetic account is returned and also `None` is stored.